sql: panic on hydrating dropped types

**Describe the problem**

CockroachDB hits an NPE when trying to fetch all descriptors after dropping a database with a type and a table referencing it. In this case, the crash was manifested via a `SHOW GRANTS` command on a table.

**To Reproduce**

On Cockroach demo, run:
```
CREATE DATABASE test;
USE test;
CREATE TYPE greeting AS ENUM ('hi');
CREATE TABLE t(a greeting);
USE defaultdb;
DROP DATABASE test CASCADE;
CREATE TYPE another AS ENUM ('hi');
CREATE TABLE s(a another);
SHOW GRANTS ON s;
```

This should make Cockroach crash:
```
root@127.0.0.1:53997/test> show grants on t;
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x48 pc=0x5400dbb]

goroutine 10789 [running]:
github.com/cockroachdb/cockroach/pkg/sql/catalog/descs.HydrateGivenDescriptors.func1(0x977a8a0, 0xc003703f00, 0x3c, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/pbardea/Developer/go/src/github.com/cockroachdb/cockroach/pkg/sql/catalog/descs/collection.go:1510 +0x8b
github.com/cockroachdb/cockroach/pkg/sql/catalog/typedesc.TypeLookupFunc.GetTypeDescriptor(0xc001ea15c0, 0x977a8a0, 0xc003703f00, 0x3c, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/pbardea/Developer/go/src/github.com/cockroachdb/cockroach/pkg/sql/catalog/typedesc/type_desc.go:494 +0x87
github.com/cockroachdb/cockroach/pkg/sql/catalog/typedesc.HydrateTypesInTableDescriptor.func1(0xc000ea2900, 0x0, 0x0)
	/Users/pbardea/Developer/go/src/github.com/cockroachdb/cockroach/pkg/sql/catalog/typedesc/type_desc.go:528 +0x123
github.com/cockroachdb/cockroach/pkg/sql/catalog/typedesc.HydrateTypesInTableDescriptor(0x977a8a0, 0xc003703f00, 0xc002959900, 0x96e7a80, 0xc001ea15c0, 0x0, 0x0)
	/Users/pbardea/Developer/go/src/github.com/cockroachdb/cockroach/pkg/sql/catalog/typedesc/type_desc.go:540 +0xd1
github.com/cockroachdb/cockroach/pkg/sql/catalog/descs.HydrateGivenDescriptors(0x977a8a0, 0xc003703f00, 0xc002fa4840, 0x2c, 0x2c, 0x0, 0xc002fa4840)
	/Users/pbardea/Developer/go/src/github.com/cockroachdb/cockroach/pkg/sql/catalog/descs/collection.go:1529 +0x1ab
github.com/cockroachdb/cockroach/pkg/sql/catalog/descs.(*Collection).GetAllDescriptors(0xc001a54230, 0x977a8a0, 0xc003703f00, 0xc00126e630, 0x0, 0xc000437040, 0xb, 0x4069c52, 0x405739c)
	/Users/pbardea/Developer/go/src/github.com/cockroachdb/cockroach/pkg/sql/catalog/descs/collection.go:1472 +0x124
github.com/cockroachdb/cockroach/pkg/sql.forEachTableDescWithTableLookupInternal(0x977a8a0, 0xc003703f00, 0xc001a54420, 0x0, 0x0, 0x0, 0xc003c6fda0, 0xc001638e40, 0x2)
	/Users/pbardea/Developer/go/src/github.com/cockroachdb/cockroach/pkg/sql/information_schema.go:1807 +0x69
github.com/cockroachdb/cockroach/pkg/sql.forEachTableDescWithTableLookup(...)
	/Users/pbardea/Developer/go/src/github.com/cockroachdb/cockroach/pkg/sql/information_schema.go:1768
github.com/cockroachdb/cockroach/pkg/sql.forEachTableDesc(0x977a8a0, 0xc003703f00, 0xc001a54420, 0x0, 0x0, 0xc002926e00, 0x60, 0x8370060)
	/Users/pbardea/Developer/go/src/github.com/cockroachdb/cockroach/pkg/sql/information_schema.go:1697 +0x8d
github.com/cockroachdb/cockroach/pkg/sql.populateTablePrivileges(0x977a8a0, 0xc003703f00, 0xc001a54420, 0x0, 0xc001638e40, 0xc002926f80, 0x4067ac0)
	/Users/pbardea/Developer/go/src/github.com/cockroachdb/cockroach/pkg/sql/information_schema.go:1383 +0x81
github.com/cockroachdb/cockroach/pkg/sql.virtualDefEntry.getPlanInfo.func1.1(0x96e7940, 0xc00349a2c0, 0x2, 0x1)
	/Users/pbardea/Developer/go/src/github.com/cockroachdb/cockroach/pkg/sql/virtual_schema.go:492 +0x11a
github.com/cockroachdb/cockroach/pkg/sql.setupGenerator.func2(0x977a8a0, 0xc003ab4180, 0xc002a4faa0, 0xc002658cc0, 0xc00349a2c0)
	/Users/pbardea/Developer/go/src/github.com/cockroachdb/cockroach/pkg/sql/virtual_table.go:115 +0x109
created by github.com/cockroachdb/cockroach/pkg/sql.setupGenerator
	/Users/pbardea/Developer/go/src/github.com/cockroachdb/cockroach/pkg/sql/virtual_table.go:104 +0xf7
```

I only briefly looked at what's going on here, but it looks like the issue here is that when we drop a database, we drop the type descriptor immediately, but the table descriptor hangs around referencing the dropped type. When we want to create a collection with all descriptors, we try and hydrate the type which no longer exists on the dropped table descriptor.

**Expected behavior**
A clear and concise description of what you expected to happen.


**Environment:**
Dev machine. Cockroach version:
```
Build Tag:        v20.2.0-alpha.3-1051-g0c89e7b81a
Build Time:       2020/09/13 20:31:28
Distribution:     CCL
Platform:         darwin amd64 (x86_64-apple-darwin19.5.0)
Go Version:       go1.14.4
C Compiler:       4.2.1 Compatible Apple LLVM 11.0.3 (clang-1103.0.32.62)
Build Commit ID:  0c89e7b81aa65d6afa35135faf03c37a0082cb8b
Build Type:       development
```

--- a/pkg/sql/catalog/descs/collection.go
+++ b/pkg/sql/catalog/descs/collection.go
@@ -1083,13 +1083,17 @@ func (tc *Collection) resolveSchemaByID(
 // hydrateTypesInTableDesc installs user defined type metadata in all types.T
 // present in the input TableDescriptor. It always returns the same type of
 // TableDescriptor that was passed in. It ensures that ImmutableTableDescriptors
-// are not modified during the process of metadata installation.
+// are not modified during the process of metadata installation. Dropped tables
+// do not get hydrated.
 //
 // TODO(ajwerner): This should accept flags to indicate whether we can resolve
 // offline descriptors.
 func (tc *Collection) hydrateTypesInTableDesc(
 	ctx context.Context, txn *kv.Txn, desc catalog.TableDescriptor,
 ) (catalog.TableDescriptor, error) {
+	if desc.Dropped() {
+		return desc, nil
+	}
 	switch t := desc.(type) {
 	case *tabledesc.Mutable:
 		// It is safe to hydrate directly into Mutable since it is
@@ -1602,6 +1606,10 @@ func HydrateGivenDescriptors(ctx context.Context, descs []catalog.Descriptor) er
 		// Now hydrate all table descriptors.
 		for i := range descs {
 			desc := descs[i]
+			// Never hydrate dropped descriptors.
+			if desc.Dropped() {
+				continue
+			}
 			if tblDesc, ok := desc.(*tabledesc.Immutable); ok {
 				if err := typedesc.HydrateTypesInTableDescriptor(
 					ctx,

--- a/pkg/sql/logictest/testdata/logic_test/enums
+++ b/pkg/sql/logictest/testdata/logic_test/enums
@@ -1202,3 +1202,21 @@ INSERT INTO table_with_not_null_enum_no_vals VALUES (1);
 statement ok
 ROLLBACK; DROP TABLE table_with_not_null_enum_no_vals; DROP TYPE enum_with_no_vals;
 
+# Regression test that hydrating descriptors does not happen on dropped
+# descriptors. See #54343.
+
+subtest dropped_database_with_enum
+
+statement ok
+CREATE DATABASE to_drop;
+USE to_drop;
+CREATE TYPE greeting AS ENUM ('hi');
+CREATE TABLE t(a greeting);
+USE defaultdb;
+DROP DATABASE to_drop CASCADE;
+
+# Before the bug-fix which introduced this test, this call would load all
+# descriptors, including dropped ones, and hydrate them, causing a panic as
+# the referenced type no longer exists.
+statement ok
+SELECT * FROM crdb_internal.tables;

